src/lsm/compaction.zig

```diff
@@ -250,22 +253,33 @@ pub fn CompactionType(
             assert(compaction.filter.state == .building);
             assert(compaction.data.state == .building);
 
-            // TODO Implement manifest.move_table() optimization if there's only range.table_count == 1.
-            // This would do update_tables + insert_tables inline without going through the iterators.
-
-            const iterator_b_context = .{
-                .grid = grid,
-                .manifest = manifest,
-                .level = level_b,
-                .snapshot = op_min,
-                .key_min = range.key_min,
-                .key_max = range.key_max,
-                .direction = .ascending,
-                .table_info_callback = iterator_b_table_info_callback,
-            };
+            // Start the iteraters if we're not only moving one table.
+            if (!compaction.move_table()) {
+                const iterator_b_context = .{
+                    .grid = grid,
+                    .manifest = manifest,
+                    .level = level_b,
+                    .snapshot = op_min,
+                    .key_min = range.key_min,
+                    .key_max = range.key_max,
+                    .direction = .ascending,
+                    .table_info_callback = iterator_b_table_info_callback,
+                };
+
+                compaction.iterator_a.start(iterator_a_context, iterator_a_io_callback);
+                compaction.iterator_b.start(iterator_b_context, iterator_b_io_callback);
+            }
+        }
 
-            compaction.iterator_a.start(iterator_a_context, iterator_a_io_callback);
-            compaction.iterator_b.start(iterator_b_context, iterator_b_io_callback);
+        /// Returns true if the compaction must use the move table optimization. 
```

Comment:
There is another complication to move-table that is not yet implemented.

Consider the case of "append-only" trees (i.e. insert-only trees which are keyed by `timestamp`, such as the `Transfer` object table). If we always append keys, then tables never overlap. And if tables never overlap, then we always use move table.

The problem with that is: those tables may be any size. If they are quite small, then the space amplification is high. There are only a limited number of tables too, because of metadata overhead.

The solution is at the last level to coalesce adjacent (but nonoverlapping) tables (instead of a simple move-table).

Reply:
> The solution is at the last level to coalesce adjacent (but nonoverlapping) tables (instead of a simple move-table).

Actually we probably want to be more aggressive than just doing this for the last level. Maybe based on how full the tables are instead.

Reply:
Would the coalescing decision be done in `Manifest.compaction_table`? I assume it would return a range of multiple tables (`range.table_count > 1`) by defaulting not hitting the move-table Compaction paths.

Reply:
Yeah, that sounds good.
Btw, we can include "number of values in the table" to the `TableInfo` so that we know when tables can be coalesced.
(Also keep in mind that the table size can differ between levels, see https://github.com/tigerbeetledb/tigerbeetle/issues/486).

Reply:
> Actually we probably want to be more aggressive than just doing this for the last level. Maybe based on how full the tables are instead.

Let's hold off on coalescing middle levels for now.

For the last level, coalescing converts a table of normal size (according to the tree) to a table of much larger size (8x bigger), but this is fine since the last level doesn't compact into further levels. Making the table bigger like this doesn't increase the number tables involved in subsequent compactions, since there aren't any for the last level. This is also enough to solve our metadata overhead problem, because the last level has ~90% of the data.

However, for interim levels, of course we can't coalesce to make the table bigger than the normal table size, since this would increase the number of overlapping tables involved in subsequent tables, and vary compaction latencies.

Where coalescing would help for interim levels, is where tables are not being saturated as we expect, up to their normal table size. But I think let's first explore this and get a feeling for how many tables in interim levels are grossly underweight.

Coalescing within a level, ideally also requires the presence of an adjacent table, where both tables together are still within the normal table size. It would be good to get a feeling for how many underweight tables we have typically side by side, if this optimization would pay off. Otherwise, if we coalesce an underweight table with a full weight table, do we end up in a better position?

Reply:
> However, for interim levels, of course we can't coalesce to make the table bigger than the normal table size, since this would increase the number of overlapping tables involved in subsequent tables, and vary compaction latencies.

For object trees (which are append-only) and low-cardinality secondary indexes (which have a handful of long runs of strictly increasing keys because of the timestamp suffix), it doesn't increase the number of overlapping tables, because tables never / rarely overlap (respectively).

> Where coalescing would help for interim levels, is where tables are not being saturated as we expect, up to their normal table size.

@jamii measured this yesterday and it impacts almost all tables, which is specifically why I think we need to coalesce before reaching the last level. Otherwise we are paying a massive space amplification cost — even when data (finally) starts reaching the last level, all previous levels will always remain full of mostly-empty tables.

> Otherwise, if we coalesce an underweight table with a full weight table, do we end up in a better position?

This is why I suggest tracking the table's `value_count` in the `TableInfo` — we can tell in advance whether or not coalescing is an improvement, and use that to decide whether to trigger "coalesce-compaction" or "move-table-compaction". e.g. if we have a table that is at 1/8 capacity, and it will have an adjacent (nonoverlapping) table at 1/4 capacity, we are guaranteed that the coalesced table will be at 3/8 capacity.